kinder/pkg/actions/waiter.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actions
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	kcluster "k8s.io/kubeadm/kinder/pkg/cluster"
+	"sigs.k8s.io/kind/pkg/exec"
+)
+
+// waitNewControlPlaneNodeReady waits for a new control plane node reaching the target state after init/join
+func waitNewControlPlaneNodeReady(kctx *kcluster.KContext, kn *kcluster.KNode, flags kcluster.ActionFlags) error {
+	fmt.Printf("==> Waiting for Node and control-plane pods getting Ready (timeout %s) 📦\n", flags.Wait)
+	if pass := waitFor(kctx, kn, flags.Wait,
+		nodeIsReady,
+		staticPodIsReady("kube-apiserver"),
+		staticPodIsReady("kube-controller-manager"),
+		staticPodIsReady("kube-scheduler"),
+	); !pass {
+		return errors.Errorf("timeout: Node and control-plane did not reach target state")
+	}
+	fmt.Println()
+	return nil
+}
+
+// waitControlPlaneUpgraded waits for a control plane node reaching the target state after upgrade
+func waitControlPlaneUpgraded(kctx *kcluster.KContext, kn *kcluster.KNode, flags kcluster.ActionFlags) error {
+	version := flags.UpgradeVersion.String()
+
+	fmt.Printf("==> Waiting for control-plane pods restarting with the new version (timeout %s) 📦\n", flags.Wait)
+	if pass := waitFor(kctx, kn, flags.Wait,
+		staticPodHasVersion("kube-apiserver", version),
+		staticPodHasVersion("kube-controller-manager", version),
+		staticPodHasVersion("kube-scheduler", version),
+	); !pass {
+		return errors.Errorf("timeout: control-plane did not reach target state")
+	}
+	fmt.Println()
+	return nil
+}
+
+// waitKubeletUpgraded waits for a node reaching the target state after upgrade
+func waitKubeletUpgraded(kctx *kcluster.KContext, kn *kcluster.KNode, flags kcluster.ActionFlags) error {
+	version := flags.UpgradeVersion.String()
+
+	fmt.Printf("==> Waiting for node restarting with the new version (timeout %s) 📦\n", flags.Wait)
```

Comment:
`fmt.Printf("==> Waiting for Node to restart with the new version (timeout %s) 📦\n", flags.Wait)`